### ISSUE TITLE
cirrus CI reports HTTP error 502 with message suggesting to retry

### ISSUE THREAD

**Issue by berrange** [sio/cirrus-run#7]
We saw a transient failure from Cirrus CI / cirrus-run in this libvirt job:

https://gitlab.com/libvirt/libvirt/-/jobs/704244941

The build completed and cirrus-run displayed the build log, but after that it got an HTTP502 error

```
<h1>Error: Server Error</h1>
<h2>The server encountered a temporary error and could not complete your request.<p>Please try again in 30 seconds.</h2>
```

502 is a fairly generic HTTP error code, so not an obvious code to trigger a retry on. Matching on the error message text suggesting the retry is a bit gross. Perhaps there were other HTTP headers present that could be used to indicate need to retry, but they're not visible from verbose logs.

**Comment by sio** [comment]
I think there is not much more cirrus-run could do in that case.

[A retry for non-200 HTTP codes](https://github.com/sio/cirrus-run/blob/a5c746c80dbfc8a9e8b5d0f9e1e20ed7535c6fef/cirrus_run/api.py#L60-L80) is already in place. API call was attempted three times with 2 seconds delay before that error would be allowed to surface - that means API server experienced something more serious than a momentarily hiccup. The retries are logged a little deeper than just `-v`, so they are not appearing in your log. But you should not change your CI setup to use more verbose output because the API key would appear in stdout with `-vv`. *Maybe we should log such retries on another verbosity level?*

Server's suggestion about 30 seconds delay may have been a good one but I think that increasing the retry delay for all cirrus-run users is not a good idea: if API is down for 6 seconds - who knows when (and if) will it be up again?

One thing was peculiar about this case - the timing of the error. Here is a rough timeline of events

- The build was started
- cirrus-run was waiting on the build to finish and was sending regular API queries to check build status
- API errored out three times for one of these status checks
- cirrus-run marked the build as "error" and started fetching and showing build log
- While the log was being fetched and printed to stdout, CirrusCI build has successfully finished - that's why the log is not truncated in GitLab CI. It was a narrow window but I think is was just a coincidence. Or (less probably) maybe the API server crashed while wrapping up this very job.
- After printing the log cirrus-run reported an error it caught earlier (CirrusHTTPError)

Is this the only occurence of such error in your experience? Please report if you encounter it again.

Do you have any suggestions on how cirrus-run could've handled it differently?

**Comment by berrange** [comment]
We've had a reasonable number of transient failures, but we've only fairly recently started using the "-v" option to capture debug info, so I'm not sure if this particular error is something we've seen before. We have had at least one different error previously - the #4 issue. I just wanted to report this since it is the first time we've seen this particular failure message.

I don't have any particular compelling suggestions aside from further retries, but as you say, that would possibly have its own negative consequences. May be a cli parameter for tuning number of retries or the retry interval ? I'm fine if we just leave this issue though, waiting to see if it re-ocurrs at all often.

**Comment by sio** [comment]
Yeah, let's leave this open and see if it reoccurs. Maybe we'll tweak the retry parameters then.

Thank you very much for reporting! It seems that all the interesting stuff happens to you first even though I run more jobs in my projects :) Maybe it's related to being in different timezone - most of my jobs run when USA is asleep

**Comment by sio** [comment]
I've added some [automated tests](https://github.com/sio/cirrus-run/blob/master/tests/test_api_retries.py) which prove that retries work as expected.

**Comment by sio** [comment]
Cirrus API server has become really flaky lately, so I've added [a workaround](https://github.com/sio/cirrus-run/commit/677b3beaf08438a6f4c28707e3758e56206a8e58) for this specific error message. A one-time delay of 30 seconds will be added if GraphQL API server spits out HTTP 50* with this specific text. Fix is included in v0.4.0, already published at PyPI

**Comment by sio** [comment]
I've just published v0.5.0 which further enhances handling of intermittent API server errors.
